include/gz/math/TimeVaryingVolumetricGridLookupField.hh

```diff
@@ -0,0 +1,291 @@
+        {
+          newSess.iter = nextTime;
+          nextTime = std::next(nextTime);
+        }
+        newSess.time = _time;
+        return newSess;
+      }
+
+      /// \brief Documentation inherited
+      public: std::vector<InterpolationPoint4D<T, V>>
+        LookUp(const InMemorySession<T, V> &_session,
+          const Vector3<V> &_point,
+          const Vector3<V> &_tol = Vector3<V>{1e-6, 1e-6, 1e-6}) const {
+
+        std::vector<InterpolationPoint4D<T, V>> res;
+
+        if (_session.iter == this->gridFields.end())
+        {
+          // Out of bounds
+          return res;
+        }
+
+        InterpolationPoint4D<T, V> slice1;
+        slice1.timeSlice = _session.iter->second.GetInterpolators(
+            _point, _tol.X(), _tol.Y(), _tol.Z());
+        slice1.time = _session.iter->first;
+        res.push_back(slice1);
+
+        auto nextTime = std::next(_session.iter);
+        if (nextTime != this->gridFields.end())
+        {
+          // Only add next item if field exists
+          InterpolationPoint4D<T, V> slice2;
+          slice2.timeSlice = nextTime->second.GetInterpolators(
+            _point, _tol.X(), _tol.Y(), _tol.Z());
+          slice2.time = nextTime->first;
+          res.push_back(slice2);
+        }
+        return res;
+      }
+
+      /// \brief Uses quadrilinear interpolation to estimate value of current
+      /// point. Returns nullopt if query is out of range.
+      /// \param[in] _session - The session
+      /// \param[in] _interpolators - The interpolation points
+      /// retrieved from `LookUp()`
+      /// \param[in] _values1 - Value array at timestep 1.
+      /// \param[in] _values2 - Value array at timestep 2.
+      /// \param[in] _position - The position to be queried.
+      /// \param[in] _default - Value used if there is a hole in the data.
+      /// \returns The estimated value for the point. Nullopt if we are
+      /// outside the field. Default value if in the field but no value is
+      /// in the index.
+      public: template<typename X>
+      std::optional<X> EstimateQuadrilinear(
+        const InMemorySession<T, V> &_session,
+        const std::vector<InterpolationPoint4D<T, V>> &_interpolators,
+        const std::vector<X> &_values1,
+        const std::vector<X> &_values2,
+        const Vector3<X> &_position,
+        const X _default = X(0)
+      ) const
+      {
+        if (_session.iter == this->gridFields.end())
+        {
+          // Out of bounds
+          return std::nullopt;
+        }
+
+        auto time = _session.time;
+        if (_interpolators.size() == 0) return std::nullopt;
+        if (_interpolators.size() == 1)
+        {
+          // This happens we reach the end of time
+          return _session.iter->second.EstimateValueUsingTrilinear(
+            _position,
+            _values2,
+            _default);
+        }
+
+        /// Got nothing to interpolate. Out of bounds.
+        if (_interpolators[0].timeSlice.size() == 0
```

Comment:
@arjo129 FYI, safe to ignore: there's [`std::vector<T, Allocator>::empty`](https://en.cppreference.com/w/cpp/container/vector/empty) too.